### ISSUE TITLE
bug: should invalidate some cache when scaling

### ISSUE THREAD

**Issue by BugenZhao** [risingwavelabs/risingwave#5567]
Under current design of scaling-out, we'll keep the original actors untouched and only create new ones for new partitions. We assume the operator cache of original ones will be evicted gradually as the system running with new configuration.

However, this might be problematic if we then scale-in: some partition may be assigned back to the actors, but the entries that are not evicted yet will be stale. This will lead to wrong results!

We need to evict some cache entries that not belongs to the old partitions when scaling. Note that the cache of each operator is implemented individually, this requires the operator itself realize the partition strategy of consistent hash.

**Comment by jon-chuang** [comment]
Can we refactor all of the hashmap cache to compute a hash that contains the consistent hash? So we don't compute the hash multiple times...

**Comment by BugenZhao** [comment]
> Can we refactor all of the hashmap cache to compute a hash that contains the consistent hash? So we don't compute multiple different hashes

Yes, it might be beneficial to pre-compute the consistent hash and store it. However, we should avoid using the same hash strategy as explained in #5579. 🥺

**Comment by jon-chuang** [comment]
So, we might be able to easily clean the via a prefix scan of the hashmap. So we can change it to a BtreeMap<HashOfKey, Value>, and scan the hash for the first k bits which are the vnodes.

**Comment by BugenZhao** [comment]
I decide not to dive into this issue deeply, at least for now. Considering that there's no frequent scaling in the system, it's okay to simply drop all cache entries if we're not sure about the cache freshness. Or specifically, when the new vnode bitmap is not a subset of the old one. So for scaling out, this will not be triggered and everything performs well like before.

**Comment by fuyufjh** [comment]
> I decide not to dive into this issue deeply, at least for now. Considering that there's no frequent scaling in the system, it's okay to simply drop all cache entries if we're not sure about the cache freshness. Or specifically, when the new vnode bitmap is not a subset of the old one.
> 
> So for scaling out, this will not be triggered and everything performs well like before. For scaling in, which will not be a critical path, I think the performance loss is tolerable.
> 
> cc @fuyufjh

Sounds sensible to me. We can record this as a known issue for now.

**Comment by BugenZhao** [comment]
> So we can change it to a BtreeMap<HashOfKey, Value>, and scan the hash for the first k bits which are the vnodes.

That is a very interesting idea. 🤩 But this requires us to use the exact same hash function for exchange and grouping operators, which may prevent some (further) optimization (like the `HashKey` we use now?). A more trivial equivalent could be `BTreeMap<Vnode, Cache>` 🤣 which sounds reasonable in dynamic filter?

**Comment by jon-chuang** [comment]
Actually, another option is simply to compute vnode upon read when `cache_may_be_stale == true`, and drop the entry otherwise. 

Personally, I think the cost of deleting the entire cache could be expensive and the repopulation will be costly and cause latency spikes. 

If the cache size is small, a one-time scan-and-delete (compute vnode for each row) is also an option. 

Lastly, as an advanced in-between solution, poll a background future which does scan and delete of a certain number of entries per epoch and remembers the range it has scanned so far, and which performs a staleness check upon read until the full scan is complete.

**Comment by BugenZhao** [comment]
> Actually, another option is simply to compute vnode upon read when `cache_may_be_stale == true`, and drop the entry otherwise.
> 
> Personally, I think the cost of deleting the entire cache could be expensive and the repopulation may be costly and cause latency spikes.

Agreed. Just for the sake of correctness, we merge #5813 as a temporary solution and leave this issue open for further discussion.

> Lastly, as an advanced in-between solution, poll a background future...

Yes. I've also considered a similar idea to introduce a "configuration version" for each entry (or simply use the epoch in the LRU cache), and refuse to give stale results but evict it automatically and lazily when it is accessed again.

> We only need to intercept their `get` function and/or their `scan` or `range` functions and the rest can be implemented generically via trait interface.

Sounds great. Anyway, we should unify the caches in different operators first (by struct or by trait, both ok).

**Comment by jon-chuang** [comment]
I think one other issue is during recovery, we don’t have mvcc caches, this actually forces entire cluster to recover with 0 cache.
